### PR TITLE
Tablet ui tablet is overlay

### PR DESCRIPTION
**Pull request by sethalves**
- allow tablet-ui to optionally be an overlay rather than a model so that it's not visible to others

### CODE REVIEW

**Comment by sethalves** [comment]
to test:
- use the tablet-ui in various circumstances
- use general settings to switch the tablet between being visible to others or not
- make sure others see it or not
- make sure the tablet works for the owner, either way

**Comment by hifi-gustavo** [comment]
Build console output for commit: a286c72b3bf20ee67d25604f33ceb734da4749f5 available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-a286c72.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-a286c72.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0a2a1b1637ddc507531656df2549a40c8b2ea71e available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-0a2a1b1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-0a2a1b1.dmg

**Comment by sethalves** [comment]
test this please

**Comment by hifi-gustavo** [comment]
Build console output for commit: 0a2a1b1637ddc507531656df2549a40c8b2ea71e available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-0a2a1b1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-0a2a1b1.dmg

**Comment by PhilipRosedale** [comment]
problems I see so far:  

1.  Tablet QML surface comes up without the associated 'frame' model, if content is still loading.  

2.  grabbing with one hand and then clicking with the other got me in a state where I seem to have equipped the tablet to the clicking hand.  

3.  Bringing up the tablet then walking away caused the tablet to switch to the 'privacy' version.

**Comment by PhilipRosedale** [comment]
And:  

Still too slow:  on pressing the menu button, it takes >0.5sec for the tablet to appear.  

and on using stylus to click buttons - still very possible to miss/not activate.  Which has gotta be that scrolling problem.   Maybe a virtual tablet where your contact point can pass through just doesn't work for both click and scroll/drag.  hmm.

**Comment by PhilipRosedale** [comment]
also, the first time you bring up the menu or audio menu... there is too big a delay.

**Comment by hifi-gustavo** [comment]
Build console output for commit: eedc5c1647cc9f8095fde848a24b9e9e7b46842c available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-eedc5c1.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-eedc5c1.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: b2a35f45db675837dd5e20c991a853e94d76f91d available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-b2a35f4.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-b2a35f4.dmg

**Comment by sethalves** [comment]
I haven't fixed the slowness-related issues, but I think I've fixed the others.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1287beb3a2bcf6bca4a88233aac323d91c041302 available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-1287beb.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-1287beb.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 36bcaa8d0fdbca70e91566484f3c73e2145dafae available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-36bcaa8.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-36bcaa8.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1864975c51bad5fb79423c686d38ae513acbf9db available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-1864975.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-1864975.dmg

File: interface/src/Application.cpp [hunk 1]
```diff
@@ -781,6 +783,11 @@ Application::Application(int& argc, char** argv, QElapsedTimer& startupTimer, bo
     connect(&domainHandler, SIGNAL(connectedToDomain(const QString&)), SLOT(updateWindowTitle()));
     connect(&domainHandler, SIGNAL(disconnectedFromDomain()), SLOT(updateWindowTitle()));
     connect(&domainHandler, SIGNAL(disconnectedFromDomain()), SLOT(clearDomainOctreeDetails()));
+    connect(&domainHandler, &DomainHandler::disconnectedFromDomain, this, [this]() {
+        getOverlays().deleteOverlay(getTabletScreenID());
+        getOverlays().deleteOverlay(getTabletHomeButtonID());
+        getOverlays().deleteOverlay(getTabletFrameID());
+    });
```
Comment (hyperlogic) [inline interface/src/Application.cpp]: I thought overlays were auto deleted when you switch domains.

File: interface/src/Application.h [hunk 2]
```diff
@@ -672,6 +676,8 @@ private slots:
     void addAssetToWorldInfoDone(QString modelName);
     void addAssetToWorldError(QString modelName, QString errorText);
 
+    bool _mouseToOverlays { false };
```
Comment (hyperlogic) [inline interface/src/Application.h]: IMHO, I don't like that this is a member variable, can't you just make it an argument that is passed to all the MouseEventHandlers?

File: interface/src/ui/overlays/Overlays.cpp [hunk 3]
```diff
@@ -719,11 +728,41 @@ PointerEvent Overlays::calculatePointerEvent(Overlay::Pointer overlay, PickRay r
...
+        return rayPickResult;
+    }
+    // then tablet frame
+    overlaysToInclude.clear();
+    overlaysToInclude << OverlayID(qApp->getTabletFrameID());
+    rayPickResult = findRayIntersectionInternal(ray, true, overlaysToInclude, overlaysToDiscard);
+    if (rayPickResult.intersects) {
+        return rayPickResult;
+    }
+    // then whatever
```
Comment (hyperlogic) [inline interface/src/ui/overlays/Overlays.cpp]: I don't understand why we need this, can you explain please?  Also,  It seems wrong to hard-code the priority of the tablet, tabletFrame and tabletButton, is there a way to generalize this?

File: interface/src/Application.cpp [hunk 4]
```diff
@@ -3126,21 +3145,23 @@ void Application::mousePressEvent(QMouseEvent* event) {
         event->buttons(), event->modifiers());
 
     if (!_aboutToQuit) {
-        getOverlays().mousePressEvent(&mappedEvent);
-
-        if (!_controllerScriptingInterface->areEntityClicksCaptured()) {
+        if (getOverlays().mousePressEvent(&mappedEvent)) {
+            _mouseToOverlays = true;
```
Comment (hyperlogic) [inline interface/src/Application.cpp]: Why do we need this?  Perhaps a comment would be helpful

File: interface/src/ui/overlays/Overlays.cpp [hunk 5]
```diff
@@ -341,27 +341,26 @@ OverlayID Overlays::getOverlayAtPoint(const glm::vec2& point) {
...
-                if (thisOverlay->findRayIntersection(origin, direction, distance, thisFace, thisSurfaceNormal)) {
-                    return thisID;
-                }
-            }
+            // auto thisOverlay = std::dynamic_pointer_cast<Base3DOverlay>(i.value());
+            // if (thisOverlay && !thisOverlay->getIgnoreRayIntersection()) {
+            //     if (thisOverlay->findRayIntersection(origin, direction, distance, thisFace, thisSurfaceNormal)) {
+            //         return thisID;
+            //     }
+            // }
```
Comment (hyperlogic) [inline interface/src/ui/overlays/Overlays.cpp]: was leaving this commented intentional?

File: interface/src/Application.cpp (hunk 1 above)
Comment (sethalves) [inline interface/src/Application.cpp]: no, overlays persist between domain changes.

File: interface/src/ui/overlays/Overlays.cpp (hunk 3 above)
Comment (sethalves) [inline interface/src/ui/overlays/Overlays.cpp]: I'm sure there is, but I'm hesitant to take on reworking our event system...

File: interface/src/ui/overlays/Overlays.cpp (hunk 5 above)
Comment (sethalves) [inline interface/src/ui/overlays/Overlays.cpp]: I'm not sure.  I was having a lot of problems with mouse-clicks intended for the tablet getting eaten by, for example, edit's light-entity overlays.  I can't find anything that breaks with this code gone, so maybe it should be gone.

**Comment by hyperlogic** [comment]
Dev approved from my phone

**Comment by hifi-gustavo** [comment]
Build console output for commit: 8110921abfd9d1e59f029c689f6a48ee0af3bca6 available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-8110921.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-8110921.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: df034ea6dab2caa30cdb5b551bcfdf07ef5d1870 available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-df034ea.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-df034ea.dmg

**Comment by Nex-Pro** [comment]
Looking at this shortly

**Comment by Nex-Pro** [comment]
use the tablet-ui in various circumstances: **DONE**  HMD and desktop with helper.
use general settings to switch the tablet between being visible to others or not: **PASS**
make sure others see it or not: **PASS**
make sure the tablet works for the owner, either way:  **PASS**

@sethalves Big thumbs up.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 117be16c31a3234bfd209fcf2c2ea039954f2ef0 available at: https://gist.github.com/87de0e8b1de21980f912e17f22f8a525<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-117be16.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR9743-117be16.dmg

**Pull request merged**